fixed git issue.  Also New-Schema

ahandley opened the pull request:
Lots of junk XML with this release because I had a minor snafu with the NuGet and our VCS.  Should be all resolved now, and then the code for New-BqSchema is here.

quoctruong commented on Google.PowerShell.IntegrationTests/packages.config:
```diff
@@ -1,10 +1,10 @@
   <package id="Newtonsoft.Json" version="7.0.1" targetFramework="net452" />
   <package id="Zlib.Portable.Signed" version="1.11.0" targetFramework="net452" />
-</packages>
+</packages>
```
Missing a line here.

quoctruong commented on Google.PowerShell.IntegrationTests/app.config:
```diff
@@ -0,0 +1,15 @@
+      </dependentAssembly>
+    </assemblyBinding>
+  </runtime>
+</configuration>
```
Missing a line here.

ahandley replied:
No idea why it isnt showing that this one is fixed.  If you view the file, the newline is there.

quoctruong commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -0,0 +1,135 @@
+                Mode = "NULLABLE";
+            }
+
+            foreach (TableFieldSchema t in InputObject.Fields)
```
Instead of foreach loop, you can do a LINQ expression like:
```if (InputObject.Fields.Any(field => Name.Equals(field.Name)))```

quoctruong replied:
Also, in this case, should we write a terminating error to the user instead of just a normal error? Since the object returned is not going to be valid anyway.

ahandley replied:
Yep.  As I was writing this I was thinking about a lambda expression kinda thing, but I wasn't confident in my C# foo.  

And about the errors, yes.  There's no point in continuing on after that, so it'll be better not to harm the original array.  I'll add something to the docs about that.

quoctruong commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -0,0 +1,135 @@
+    /// [BigQuery Tables]
+    /// </para>
+    /// </summary>
+    [Cmdlet(VerbsCommon.New, "BqSchema")]
```
I feel like this cmdlet should probably be named "Add-BqFieldSchema" instead of "New-BqSchema" (the -InputObject parameter can be named "BqSchema"). @chrsmith and @ILMTitan what do you guys think?

ahandley replied:
I think "Add-BqSchemaField" would sound better, but I agree on the switch from New- to Add-.

ILMTitan replied:
I would propose a larger change, where the cmdlet remains New-BqSchema, but it returns 'TableFieldSchema' objects. 'TableSchema' is kind of useless, being just a wrapper around a list of 'TableFieldSchema' objects. This cmdlet's `ProcessRecord()` would just pass along any InputObjects, and then `EndProcessing()` would create a new one and append it to the pipeline. Set-BqSchema would then gather all of the pipeline objects, and only make the update call in `EndProcessing()`.
Look at New-GceFirewallProtocol and Add-GceFirewall for an example of what I am thinking of.

quoctruong commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -0,0 +1,135 @@
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = true)]
+        [ValidateSet("STRING", "BYTES", "INTEGER", "INT64", "FLOAT", "FLOAT64", "BOOLEAN", "BOOL", 
```
Should we set IgnoreCase = true here too?

ahandley replied:
Yep.  Was having issues with the parameter tags and forgot to put that back in after testing.

quoctruong commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -0,0 +1,135 @@
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false)]
+        [ValidateSet("NULLABLE", "REQUIRED", "REPEATED", IgnoreCase = true)]
```
If you are going to use these values elsewhere (like in a related cmdlet), I recommend using an enum. Same for Type parameter.

ahandley replied:
Was reading into see if i could just supply an array of strings, but that would fall under the same logic.  This thread seems to think not.

**http://stackoverflow.com/questions/36656171/can-i-use-a-constant-in-the-validateset-attribute-of-a-powershell-function-param**

Similarly, there was another article that I found saying that dynamic parameters was the only way to not have to hard code values.  I can look into that, but I don't think there are going to be any other cmdlets using these params.  Once they're sent the first time, they're only accessible through viewing the table objects and hopefully shouldn't ever be touched again.

quoctruong replied:
What I meant was something like [this](https://github.com/GoogleCloudPlatform/google-cloud-powershell/blob/master/Google.PowerShell/Logging/GcLogCmdlets.cs#L26) which is used [here](https://github.com/GoogleCloudPlatform/google-cloud-powershell/blob/master/Google.PowerShell/Logging/GcLogCmdlets.cs#L260)

ahandley replied:
Oh, that makes a lot more sense.  I'll see what I can do with that.

quoctruong commented on Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqTabledata.Tests.ps1:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    It "should deny invalid modes"{
+        { New-BqSchema -Name "Title" -Type "STRING" -Mode "NotAMode" -ErrorAction Stop} | Should Throw
```
What's the error message here? Same for "invalid types" and "duplicated column names" tests.

ahandley commented:
Made changes to the above, with the one exception of the name change.  Waiting for further comment there.

ahandley commented:
Added Set-BqSchema and refactored the integration test from New- that had a TODO on it.  

PTAL specifically that new cmdlet & tests while the discussion is still happening for the naming conventions.

ILMTitan commented on Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqTabledata.Tests.ps1:
```diff
@@ -0,0 +1,87 @@
+        $test_set | Remove-BqDataset -Force
+    }
+}
+
```
Remove extra line.

ILMTitan commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -0,0 +1,223 @@
+            try
+            {
+                response = request.Execute();
+                WriteObject(response);
```
The way this is written, I think we *do* want return the table to the pipeline, but only once, the last time.

ahandley replied:
So I can't respond to the previous comment, but I agree with that new layout and it won't be hard to implement.  Should the cmdlet be called "New-BqSchemaField"?  Or is it better to keep the name the same for both even if it's slightly less accurate?  

What does everyone else think?

ILMTitan replied:
I think New-BqSchema still works. Neither of the objects in question are just `Schema` and both have Schema in the name.

ahandley commented:
Aight.  Finally got that changeset working.  It's much cleaner this way too.

PTAL.

ILMTitan commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -190,18 +185,34 @@ public class SetBqSchema : BqCmdlet
-            Table response;
+            if (Columns.Any(field => InputObject.Name.Equals(field.Name)))
+            {
+                // ATTENTION:  should this throw? (to notify users that data may have been lost)
```
I agree with throwing here.

ILMTitan commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -65,12 +66,11 @@ public class NewBqSchema : PSCmdlet
         /// </summary>
         [Parameter(Mandatory = false, ValueFromPipeline = true)]
-        public TableSchema InputObject { get; set; }
+        public TableFieldSchema InputObject { get; set; }
```
I would consider changing the name of this to `PassThruObject`.

ILMTitan commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -37,22 +37,21 @@ public enum ColumnMode
-    /// PS C:\> $table.Schema = $schema
-    /// PS C:\> $table | Set-BqTable
+    /// PS C:\> $table = $dataset_books | New-BqTable "book_info"
+    /// PS C:\> $result = New-BqSchema -Name "Author" -Type "STRING" | `
```
Fyi, ending a line with a pipe is one of the few times powershell does not need a backtick to know the statement continues on the next line.

ILMTitan commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -0,0 +1,233 @@
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false, ValueFromPipeline = true)]
+        public TableFieldSchema InputObject { get; set; }
```
I would consider renaming this to `PassThruObject`

ahandley replied:
Fair enough.  I just needed a name since the users should never be using the flag explicitly.

ILMTitan commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -0,0 +1,233 @@
+        {
+            if (Columns.Any(field => InputObject.Name.Equals(field.Name)))
+            {
+                // ATTENTION:  should this throw? (to notify users that data may have been lost)
```
I agree with throwing here.

ILMTitan commented on Google.PowerShell/BigQuery/BqTabledata.cs:
```diff
@@ -0,0 +1,233 @@
+    /// <example>
+    ///   <code>
+    /// PS C:\> $table = $dataset_books | New-BqTable "book_info"
+    /// PS C:\> $result = New-BqSchema -Name "Author" -Type "STRING" | `
```
FYI, ending a line with a pipe is one of the few way to indicate the statement continues on the next line without needing a backtick.

ahandley replied:
Oh cool.  That makes things a lot cleaner.
